### ISSUE TITLE
try running the MATLAB code on linux

### ISSUE THREAD

**Issue by niryariv** [Gizra/IBLI#5]
question mark since we need to see if a linux server is feasible, based on whether we can run Anton's IDL code on GDL (#3)

**Comment by niryariv** [comment]
@alanorth I guess our first goal would be to run matlab on the server from the command line - seems like this should be possible: http://www.mathworks.com/help/matlab/matlab_env/start-matlab-on-linux-platforms.html

step 2 would be to see that the current scripts run and produce the expected output - then we can move on to automating Anton's IDL toolchain. Sounds ok?

**Comment by alanorth** [comment]
Sounds ok.  I have a copy of Matlab and I will see if I can run a tutorial or something to figure out what's possible; ie calling it on a server and getting it to spit out some numbers.

**Comment by alanorth** [comment]
Ok, running Matlab on a headless Linux server is really easy.  Given test script `test.m`:

``` matlab
sqrt(9)
quit
```

... I can do:

``` console
matlab -nojvm -nodisplay -nosplash -nodesktop -r test
```

The output is kinda noisy, but the next logical step is to use file handles inside the script anyways, for example:

``` matlab
ans = sqrt(3);

outfile = sprintf('outfile.tmp');
fh = fopen(outfile, 'w');
fprintf(fh, '%d', ans);
fclose(fh);

quit
```

With the output being more programmatically named, maybe a timestamp etc.

**Comment by niryariv** [comment]
awesome. next step is #2 then, verifying the current Kenya code runs well

**Comment by bnedavies** [comment]
@apurbashee Can you send any Kenya code as is, just to test?  Does not need to be perfect.

**Comment by niryariv** [comment]
@alanorth can you try running this code from Josh on the Linux server?  https://github.com/Gizra/IBLI/tree/master/matlab

**Comment by woodardjoshua** [comment]
Nir,

I am packaging this up now and will upload today. Thanks for your patience.

Josh

On Thu, Mar 27, 2014 at 8:17 AM, Nir Yariv notifications@github.com wrote:

> @alanorth https://github.com/alanorth can you try running this code
> from Josh on the Linux server?
> https://github.com/Gizra/IBLI/tree/master/matlab
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/Gizra/IBLI/issues/5#issuecomment-38795597
> .

## 

---

Joshua D. Woodard, PhD
Assistant Professor of Agribusiness and Finance
Zaitz Family Sesquicentenntial Faculty Fellow
David R. Atkinson Center for a Sustainable Future Faculty Fellow
Dyson School of Applied Economics and Management
Cornell University
Phone: (217) 979-1251 Fax: (815) 328-2252
Email: woodardjoshua@gmail.com, jdw277@cornell.edu
Webpage: http://dyson.cornell.edu/people/profiles/woodard.php

---

Confidentiality Notice: This message, including any attachments, is for the
sole use of the intended recipient(s) and may contain confidential and
privileged information. Any unauthorized review, use, disclosure, or
distribution is prohibited.  This email is considered a draft of material
that could be published at a later date, and should not be forwarded,
cited, or quoted without permission, and is subject to revision. If you are
not the intended recipient, please contact the sender by reply e-mail and
destroy all copies of the original message. We do not accept legal
responsibility or otherwise for the contents of this message.

**Comment by niryariv** [comment]
Thanks Josh! 

@alanorth fyi ^^ let's do a test run of that code on Linux once it's up

**Comment by woodardjoshua** [comment]
FYI, Nir and Apurba have this attachments (could not be sent over the list).

Get in touch if you run into any issues.

Thanks,

Josh

On Thu, Mar 27, 2014 at 12:03 PM, Joshua Woodard woodardjoshua@gmail.comwrote:

> Nir,
> 
> Please find attached a new set of matlab files attached (scrap the old
> ones).  This will generate rates for the LRLD season (let's get this
> working first).
> 
> I have it set up right now so that all you should need to do is open the
> file called genLRLDrate.m and you will see in the first 3 lines:
> 
> %% Code to estimate LRLD IBLI Rates
> %% Import NDVI Files
> filepathMain='C:\Users\JoshAsus2\Dropbox\IBLIMatlab\';
> 
> Change the 3rd line to your file path where you will put this folder on
> your server. If you have the right installation, you should just be able to
> run the code (Press F5 in matlab) and it should then run the models to
> estimate the base structures, run the rating procedure, and output the
> rates for the 10% and 15% products in .csv files in the same folder.
> 
> There is also a folder in the main folder called 'MatlabFunctionFiles
> which you don't need to do anything to, but which the main program will
> access. As long as they are in this folder then the code is setup such that
> matlab should be able to find them.
> 
> After we get this done, I will add the imports so that we can automate the
> updating.  For this, can you please let  me know how you have the files
> structured (filesnames etc.) for the outputs from your IDL program? Anton
> has sent these to me in a mix of ways in the past since we were just doing
> manually, but I can build the imports and match them up if I know what you
> all are calling them (filename setup for dates, etc.). In fact, easiest
> just to send over the actual files from a test run and then I can do it
> pretty quickly.
> 
> Let me know if you run into any issues. If you do, it will likely be
> something with your installation and maybe a missing toolbox or something,
> but we will see.
> 
> After we get all this running, I will send you the same files for SRSD and
> also functions for generating the index value conditional on one outcome of
> NDVI (note these are for rates/prices themselves...there will be another
> file for generating indemnities/fitted index value in a specific case).
> 
> Thanks,
> 
> Josh
> 
> On Thu, Mar 27, 2014 at 11:31 AM, Nir Yariv notifications@github.comwrote:
> 
> > Thanks Josh!
> > 
> > @alanorth https://github.com/alanorth fyi ^^ let's do a test run of
> > that code on Linux once it's up
> > 
> > ## 
> > 
> > Reply to this email directly or view it on GitHubhttps://github.com/Gizra/IBLI/issues/5#issuecomment-38818812
> > .
> 
> ## 
> 
> ---
> 
> Joshua D. Woodard, PhD
> Assistant Professor of Agribusiness and Finance
> Zaitz Family Sesquicentenntial Faculty Fellow
>  David R. Atkinson Center for a Sustainable Future Faculty Fellow
> Dyson School of Applied Economics and Management
> Cornell University
> Phone: (217) 979-1251 Fax: (815) 328-2252
> Email: woodardjoshua@gmail.com, jdw277@cornell.edu
> Webpage: http://dyson.cornell.edu/people/profiles/woodard.php
> 
> ---
> 
> Confidentiality Notice: This message, including any attachments, is for
> the sole use of the intended recipient(s) and may contain confidential and
> privileged information. Any unauthorized review, use, disclosure, or
> distribution is prohibited.  This email is considered a draft of material
> that could be published at a later date, and should not be forwarded,
> cited, or quoted without permission, and is subject to revision. If you are
> not the intended recipient, please contact the sender by reply e-mail and
> destroy all copies of the original message. We do not accept legal
> responsibility or otherwise for the contents of this message.

## 

---

Joshua D. Woodard, PhD
Assistant Professor of Agribusiness and Finance
Zaitz Family Sesquicentenntial Faculty Fellow
David R. Atkinson Center for a Sustainable Future Faculty Fellow
Dyson School of Applied Economics and Management
Cornell University
Phone: (217) 979-1251 Fax: (815) 328-2252
Email: woodardjoshua@gmail.com, jdw277@cornell.edu
Webpage: http://dyson.cornell.edu/people/profiles/woodard.php

---

Confidentiality Notice: This message, including any attachments, is for the
sole use of the intended recipient(s) and may contain confidential and
privileged information. Any unauthorized review, use, disclosure, or
distribution is prohibited.  This email is considered a draft of material
that could be published at a later date, and should not be forwarded,
cited, or quoted without permission, and is subject to revision. If you are
not the intended recipient, please contact the sender by reply e-mail and
destroy all copies of the original message. We do not accept legal
responsibility or otherwise for the contents of this message.

**Comment by niryariv** [comment]
new Matlab code here: https://github.com/Gizra/IBLI/tree/master/IBLIMatlab

**Comment by niryariv** [comment]
@alanorth did you get a chance to try running the code above on the server?

**Comment by alanorth** [comment]
@niryariv no I haven't.  Do we have an idea about how much disk space we will need yet?  I haven't provisioned the VM yet, and that's kinda important.  A "normal", low-usage Linux + Apache + PHP stack would be fine on 30GB disk, but I'm not sure if we need to store/process a bit of data ...

**Comment by niryariv** [comment]
Some calculations off the top of my head, @apurbashee @vrieling @woodardjoshua please feel free to correct / add:

The NDVI tiff is 190MB unzipped. We download it every 10 days and I understand from Anton we need to keep a backlog:

200 mb x 37 times/year = 7400 mb

Numerical data is probably much smaller, so let's say 10GB / year total and assume 5 years + some padding => say 60-80GB disk?

**Comment by vrieling** [comment]
An individual tiff-file (unzipped) is about 185 MB and we need 36 images per year (even if 72 are offered but we do not use the odd period numbers). This would mean: 6660 mb \* 13 years = 86580 mb. However, there is no need to keep these files unzipped (done in IDL program). My zipped data is now 65 GB for the whole time series (but this is because for the month-files also the odd period numbers are included).

After opening them in IDL and saving only the Kenya window (I have not heard if and how you would want to deal with Ethiopia!), an individual binary data-file is about 14 MB. The total archive is only 6.5 GB. Note that for current set-up of the IDL-programs, you need to whole archive... (perhaps could be reprogrammed by Michael?)

**Comment by niryariv** [comment]
so let's say ~120GB for existing + new data over the next 5 years (assuming the new data will be zipped periodically too) - sounds right?

@alanorth will it be possible to allocate more disk space down the road, or is that not recommended?

**Comment by vrieling** [comment]
Yes, 120 GB should do, at least for the eMODIS processing chain.

**Comment by alanorth** [comment]
Thanks for the number crunching, @niryariv and @vrieling. 120GB is starting to sound like a more serious commitment for my VM infrastructure than I initially thought (our current VMs are small, lightweight LAMP stacks!).  I really don't know if I want to commit to something so heavy (this is quite a bit of space, especially when you start thinking about snapshotted backups, etc)... let me look at my infrastructure and get back to you.

**Comment by vrieling** [comment]
@alanorth In fact, once the zipped tif's are opened, subset, and written as a binary file, you do not really need them anymore (unless you want to change the subset, let's say because you want to move to Ethiopia now...). With the zipped TIF's removed after processing, you could do for Kenya with 10-15 GB for the eMODIS processing, including at least another 5 years in future.

@apurbashee Not yet clear to me: do you eventually want to expand the current operational processing chain for eMODIS to Ethiopia?

**Comment by niryariv** [comment]
@vrieling would be feasible at a later stage to store the archived files externally (eg on an S3 bucket) and just read them when needed?

**Comment by vrieling** [comment]
@niryariv Sure why not: the archive tif's are not key input to the operational processing chain (i.e. once the subsets have been made). I don't know what an S3 bucket is, but it could be a good idea to store them on an external disk (not part of the VM). Alternatively they can always be downloaded again from: http://earlywarning.usgs.gov/fews/downloads/ (although that may take time, and could be easier to have them on disk).

**Comment by vrieling** [comment]
@alanorth @niryariv Although one thing I should mention here. The 10GB would be for the subsetted DATA (stored as individual binary files for each 10-day period). However, when processing (with current processing chain), I create also a single BIL-file containing all data (so another 10GB), and for calculating the z-scores this will create another large file (4 times as large, because these are floating points), so creating another 40GB of data. When adding some other intermediate results (averages, standard deviations), it seems then that for the current processing chain for Kenya still at least 70GB is needed... (not considering the tiffs).

**Comment by apurbashee** [comment]
@vrieling Yes we plan to expand the eMODIS processing chain to Ethiopia. Presently we are using MODIS for Ethiopia and we have a data processing program. We think moving to eMODIS will be better option since we can have uniform processing chain for both the areas and  eMODIS is filtered. Pl let me know if you anticipate any issue.

**Comment by vrieling** [comment]
@apurbashee But within the current contract with Gizra, or at some later stage? I currently set the programs up to run for Kenya. This includes sub-setting for the Kenya window, and having a raster-file with administrative IDs for Kenya. If you want to include Ethiopia, then these things will change of course: also the question arises if it should all be part of a single processing chain (treating Ethiopia/Kenya in the same way, and processing all data at once) or if this should be two separate chains (possibly combining still the subsetting part for Kenya/Ethiopia). 

Anyways, technically not a major issue, but it signifies work and time investment, so the question arises who will do it and when (now, or some time in future). This issue could probably also affect @alanorth regarding the disk sizes....

**Comment by woodardjoshua** [comment]
For what it's worth, may be wise to ser it up so that kenya/Ethiopia is
processed as one together as we are going that direction anyways. If I
understand Anton, the filtering for kenya will be a little different if we
do it with or without Ethiopia data, but presumably not to much and we will
just update policy to reflect that. In any case, may make sense to do it
now while we are at it as opposed to opening it back up later.

**Comment by vrieling** [comment]
No differences for filtering, but the data subset will be different (the window needs to include the relevant part of Ethiopia). In addition, we would probably need a single file (first shapefile that I then convert to raster to be used within my code) with unique ID-codes (so Ethiopia adminID are different from Kenyan adminID).
@woodardjoshua So for Ethiopia the same LRLD and SRSD definitions are used?
@apurbashee Please coordinate and decide how to move forward with this. Everything that I programmed so far (and transferred to Gizra) is done for Kenya only. I am working with @niryariv and Michael on this. Michael is an IDL-programmer and perhaps he can provide a useful contribution in incorporating the relevant part of Ethiopia from this point onwards in the processing chain. If done, it will have implications for disk sizes, important for @alanorth .

**Comment by apurbashee** [comment]
@vrieling I understand that it requires time and effort. Let me get back you on this..
Regarding LRLD and SRSR all definitions and processing are same for both the areas.

**Comment by alanorth** [comment]
Yeah, I'm thinking that we need to try some sort of hybrid combination of storage / archival, possibly using Amazon S3 to keep the large historical stuff off the server (if we really need it?!).

Our VM infrastructure is only really set up to host lightweight VMs.  One of the issues is keeping backups of the VM; when you have a VM which is 120 GB, then backups of the VM become 240, 360 etc and things escalate very quickly.  If we can make some compromises about the type of backups we make, like only backing up the database / code locally, then it will be much easier to manage.

**Comment by niryariv** [comment]
we probably won't have to backup the NDVI files, since these are always downloadable from the USGS site.

**Comment by alanorth** [comment]
Ok, then maybe this is a good compromise:
- Backup the MySQL database
- Backup the code
- Backup the machine's configuration (`/etc/` and friends)

But then I do not backup the VM image itself on the host.

**Comment by niryariv** [comment]
sounds good. 

(backing up the code isn't critical, since it'll be on git - but it's relatively minor in size either way)

**Comment by vrieling** [comment]
A few points:
- I spoke with @apurbashee, and we will add the 10 Ethiopian woredas to the same processing chain. This requires a few adaptations (to subset etc), but @alanorth: image sizes are not changing due to this.
- I have already made a single shapefile (containing Kenya+Ethiopia divisions). I can send it to @niryariv for posting on Git. I have also made a new raster-file (to be used in the IDL-processing). Currently I am trying to re-run the whole processing chain with the new subset and admin-units.
- @alanorth @niryariv: without considering manual edits (as I am doing now), running all programs one-after-another directly would probably take about 8-10 hours on my computer (with the biggest chunk of time for: eMODIS_createBIL_OP.pro). Although options exist to speed things up further probably, I do not have the time to dedicate myself to that. Anyway, what I want to say is that not many backups are needed (in fact I have none here...). What would need a back-up is the following probably:
  - raster-file with division IDs (but this is only 14MB and even 98 kb(!!) when zipped)
  - raster-file with diagnostics made during z-score calculation: although this is done during every update now, we want to have stable results and always consider the same pixels for each division. However, this is the same size as the previous. 
  - the IDL-codes 
  - All the other files can be re-made rather fast (as long as USGS is providing the same data over internet, which they are dedicated to do I believe).
